test/unit/app/controllers/detect-tokens-test.js

```diff
@@ -66,52 +69,61 @@ describe('DetectTokensController', function () {
 
     await controller.detectNewTokens()
     sandbox.assert.notCalled(stub)
   })
 
-  it('should only check and add tokens while in main network', async function () {
-    const controller = new DetectTokensController({ preferences: preferences, network: network, keyringMemStore: keyringMemStore })
+  it('should check and add tokens while on main network', async function () {
+    sandbox.useFakeTimers()
+    network.setProviderType(MAINNET)
+    const controller = new DetectTokensController({ preferences, network, keyringMemStore })
     controller.isOpen = true
     controller.isUnlocked = true
 
-    sandbox.stub(controller, 'detectTokenBalance')
-      .withArgs('0x0D262e5dC4A06a0F1c90cE79C7a60C09DfC884E4')
-      .returns(preferences.addToken('0x0d262e5dc4a06a0f1c90ce79c7a60c09dfc884e4', 'J8T', 8))
-      .withArgs('0xBC86727E770de68B1060C91f6BB6945c73e10388')
-      .returns(preferences.addToken('0xbc86727e770de68b1060c91f6bb6945c73e10388', 'XNK', 18))
+    const contractAddressses = Object.keys(contracts)
+    const erc20ContractAddresses = contractAddressses
+      .filter((contractAddress) => contracts[contractAddress].erc20 === true)
 
-    await controller.detectNewTokens()
-    assert.deepEqual(preferences.store.getState().tokens, [{ address: '0x0d262e5dc4a06a0f1c90ce79c7a60c09dfc884e4', decimals: 8, symbol: 'J8T' },
-      { address: '0xbc86727e770de68b1060c91f6bb6945c73e10388', decimals: 18, symbol: 'XNK' }])
-  })
+    const existingTokenAddress = erc20ContractAddresses[0]
+    const existingToken = contracts[existingTokenAddress]
+    await preferences.addToken(existingTokenAddress, existingToken.symbol, existingToken.decimals)
 
-  it('should not detect same token while in main network', async function () {
-    preferences.addToken('0x0d262e5dc4a06a0f1c90ce79c7a60c09dfc884e4', 'J8T', 8)
-    const controller = new DetectTokensController({ preferences: preferences, network: network, keyringMemStore: keyringMemStore })
-    controller.isOpen = true
-    controller.isUnlocked = true
+    const tokenAddressToAdd = erc20ContractAddresses[1]
+    const tokenToAdd = contracts[tokenAddressToAdd]
 
-    sandbox.stub(controller, 'detectTokenBalance')
-      .withArgs('0x0D262e5dC4A06a0F1c90cE79C7a60C09DfC884E4')
-      .returns(preferences.addToken('0x0d262e5dc4a06a0f1c90ce79c7a60c09dfc884e4', 'J8T', 8))
-      .withArgs('0xBC86727E770de68B1060C91f6BB6945c73e10388')
-      .returns(preferences.addToken('0xbc86727e770de68b1060c91f6bb6945c73e10388', 'XNK', 18))
+    const contractAddresssesToDetect = contractAddressses
+      .filter((address) => address !== existingTokenAddress)
+    const indexOfTokenToAdd = contractAddresssesToDetect.indexOf(tokenAddressToAdd)
+
+    const balances = new Array(contractAddresssesToDetect.length)
+    balances[indexOfTokenToAdd] = new BigNumber(10)
+
+    sandbox.stub(controller, '_getTokenBalances')
+      .returns(Promise.resolve(balances))
```

Comment:
```suggestion
      .resolves(balances)
```

Reply:
I had it this way originally - I switched it to return a Promise because it more accurately matches what the real function being stubbed out does.

It doesn't matter in this case because of an implementation detail - that we're using `await` instead of `.then`. But if we had used `.then`, this would _need_ to return a Promise

Reply:
It would probably be more correct to set a Promise return for the other cases too 🤔 The ones that shouldn't be called I mean. This is the only place it's stubbed out and expected to be called

Reply:
I'll leave it as-is for now; those test failures wouldn't be the most elegant regardless at the moment, because they'll be looking for a ~200-entry array. It kinda flooded the console when I tried testing that failure.